File: submissions/0086-partition-list/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {     
        // dummy node for holding partitioned result
        ListNode h1 = new ListNode(-1);
        ListNode tail1 = h1;
        ListNode h2 = new ListNode(-1);
        ListNode tail2 = h2;
        
        ListNode current = head;
        
        while (current != null) {
            if(current.val < x) {
                tail1.next = current;
                tail1 = current;
            } else {
                tail2.next = current;
                tail2 = current;
            }
            ListNode tmp = current;
            current = current.next;
            tmp.next = null;
        }
        
        tail1.next = h2.next;
    
        return h1.next;
    }
}

File: submissions/0086-partition-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        self.xNode = None
        self.head = head

        self.walk(head, None, x)

        return self.head
    
    def walk(self, curr: Optional[ListNode], pre: Optional[ListNode] , x: int) -> None:
        if not curr:
            return
        
        if not self.xNode:
            if curr.val >= x:
                self.preX = pre
                self.xNode = curr

            self.walk(curr.next, curr, x)
        else:
            if curr.val >= x:
                self.walk(curr.next, curr, x)
                
            else:
                nextNode = curr.next

                if not self.preX:
                    self.preX = curr
                    self.head = curr
                else:
                    self.preX.next = curr

                self.preX = curr
                curr.next = self.xNode
                    
                
                pre.next = nextNode
                
                
                self.walk(nextNode, pre, x)
        
        
